core/src/main/java/org/infinispan/interceptors/locking/ClusteringDependentLogic.java

```diff
@@ -476,15 +480,28 @@ public int getSegmentForKey(Object key) {
     */
    class ReplicationLogic extends InvalidationLogic {
       private StateTransferLock stateTransferLock;
+      private DistributionManager distributionManager;
 
       private final WriteSkewHelper.KeySpecificLogic localNodeIsPrimaryOwner = this::localNodeIsPrimaryOwner;
 
       @Inject
-      public void init(StateTransferLock stateTransferLock) {
+      public void init(StateTransferLock stateTransferLock, DistributionManager distributionManager) {
          this.stateTransferLock = stateTransferLock;
+         this.distributionManager = distributionManager;
+      }
+
+      @Override
+      public boolean localNodeIsOwner(Object key) {
+         return distributionManager.getDistributionInfo(key).write().isOwner();
       }
 
       @Override
+      public boolean localNodeIsPrimaryOwner(Object key) {
+         return distributionManager.getDistributionInfo(key).read() == Ownership.PRIMARY;
```

Comment:
> `making isPrimary() independent of reads and writes`

Have you changed that recently? In fact this was true even before... but is there any comment preferably in `CacheTopology` stating this?